### PR TITLE
Added wrapper for UPPER() and LOWER() functions for CHAR and NCHAR to handle padding for shorter strings

### PR DESCRIPTION
**Pull request by Anikait143**
## Description

### 1. Issues
- UPPER() and LOWER() functions for CHAR or NCHAR types compared to a string shorter than definition cause condition to fail.
- String functions do not handle chinese characters for NVARCHAR type when collation is not used in UPPER() and LOWER() functions

### 2. Changes made to fix the issues
- Added wrapper function handling for UPPER() and LOWER() functions to handle CHAR and NCHAR as VARCHAR and NVARCHAR respectively.
```
# Input
DECLARE @class NCHAR(2) = N'H'
SELECT '[' + @class + ']' WHERE upper(@class) = N'H'
SELECT '[' + @class + ']' WHERE lower(@class) = N'h'
```
```
# previous output (wrong output)               
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

(0 rows affected)
```
```
# current output (correct output)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[H ]                                                                                                                                                                                                                                                            

(1 rows affected)
```

```
# previous output (wrong output)
1> declare @string1 nvarchar(30) = N'比尔·拉';
2> select UPPER(@string1)
3> GO
@string1                      
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
??·?  
```
```
# current output (correct output)
1> declare @string1 nvarchar(30) = N'比尔·拉';
2> select UPPER(@string1)
3> GO
@string1                      
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
比尔·拉
```
- Added Test Cases for these issues


Task: BABEL- 3395
Authored-by: Anikait Agrawal [agraani@amazon.com](mailto:agraani@amazon.com)
Signed-off-by: Anikait Agrawal [agraani@amazon.com](mailto:agraani@amazon.com)



### Test Scenarios Covered ###
* **Use case based -**   BABEL-3395-vu-*


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -** 


* **Major version upgrade tests -** 


* **Performance tests -**


* **Tooling impact -**

* **Memory tests -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/expected/BABEL-3395-vu-prepare.out [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+~~ROW COUNT: 1~~
+
+
+-- UPPER
+CREATE VIEW dep_view_upper AS
+select UPPER(a) AS upper_a, UPPER(b) AS upper_b, UPPER(c) AS upper_c, UPPER(d) AS upper_d from upper_lower_dt WHERE UPPER(a) = N'ANIKAIT' and UPPER(b) = N'AGRAWAL' and UPPER(c) = N'ANIKAIT' and UPPER(d) = N'AGRAWAL';
+GO
+
+CREATE TABLE tab_arabic_ci_ai(col varchar(20) COLLATE arabic_ci_ai);
+INSERT INTO tab_arabic_ci_ai VALUES ('COLLATION');
```
Comment (rohit01010) [inline test/JDBC/expected/BABEL-3395-vu-prepare.out]: In such tables with column of certain collation, insert string with characters from that collation as well. For example here also insert string with arabic characters.

File: test/JDBC/expected/BABEL-3395-vu-prepare.out [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+RETURN (select TOP 1 LOWER(a) from upper_lower_dt);
+END
+GO
+
+CREATE FUNCTION dbo.tvp_func_upper_lower() 
+RETURNS TABLE 
+AS 
+RETURN 
+(
+    SELECT CAST (UPPER(a) as VARCHAR) AS upper_a, CAST (LOWER(a) as VARCHAR) AS lower_a
```
Comment (rohit01010) [inline test/JDBC/expected/BABEL-3395-vu-prepare.out]: this test will always be tested only for upgrade from 15.7 or 16.3 to latest. And since upper and lower will be fixed in 15.7 or 16.3, let's remove this CAST to `VARCHAR` on upper and lower.

File: test/JDBC/expected/BABEL-3395-vu-verify.out [hunk 3]
```diff
@@ -0,0 +1,811 @@
...
+varchar
+|??·?                          |
+~~END~~
+
+
+SELECT UPPER(NULL)
+SELECT LOWER(NULL)
+GO
+~~START~~
+text
```
Comment (rohit01010) [inline test/JDBC/expected/BABEL-3395-vu-verify.out]: `UPPER` and `LOWER` function should always return `varchar` or `nvarchar`.

File: test/JDBC/expected/BABEL-3395-vu-verify.out [hunk 4]
```diff
@@ -0,0 +1,811 @@
...
+~~START~~
+varchar
+point(1 2)
+~~END~~
+
+
+Select UPPER(a), LOWER(a) from upper_lower_text
+GO
+~~START~~
+text#!#text
```
Comment (rohit01010) [inline test/JDBC/expected/BABEL-3395-vu-verify.out]: `UPPER` and `LOWER` function should always return `varchar` or `nvarchar`.

**Review by rohit01010 — CHANGES_REQUESTED** [review]

File: test/JDBC/expected/BABEL-3395-vu-prepare.out (hunk 1 above)
Comment (Anikait143) [reply test/JDBC/expected/BABEL-3395-vu-prepare.out]: Done

File: test/JDBC/expected/BABEL-3395-vu-prepare.out (hunk 2 above)
Comment (Anikait143) [reply test/JDBC/expected/BABEL-3395-vu-prepare.out]: Done

File: test/JDBC/expected/BABEL-3395-vu-verify.out (hunk 3 above)
Comment (Anikait143) [reply test/JDBC/expected/BABEL-3395-vu-verify.out]: Done, added wrapper

File: test/JDBC/expected/BABEL-3395-vu-verify.out (hunk 4 above)
Comment (Anikait143) [reply test/JDBC/expected/BABEL-3395-vu-verify.out]: Done, Added Wrapper

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.5.0--3.6.0.sql [hunk 5]
```diff
@@ -331,6 +331,112 @@ CREATE OR REPLACE FUNCTION sys.sp_tables_internal(
...
 
+-- wrapper functions for upper --
+-- Function to handle datatypes which are implicitly convertable to VARCHAR
+CREATE OR REPLACE FUNCTION sys.upper(ANYELEMENT)
+RETURNS sys.VARCHAR
+AS $$
+DECLARE
+    varch sys.varchar;
+BEGIN
+    IF pg_typeof($1) IN ('image'::regtype, 'sql_variant'::regtype, 'xml'::regtype, 'geometry'::regtype, 'geography'::regtype) THEN
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.5.0--3.6.0.sql]: What will be behaviour for user-defined types?

Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.5.0--3.6.0.sql]: Done, Added Handling and tests

**Review by rohit01010 — APPROVED** [review]
LGTM

File: test/JDBC/expected/BABEL-3395-vu-prepare.out (hunk 1 above)
Comment (Deepesh125) [reply test/JDBC/expected/BABEL-3395-vu-prepare.out]: Have we covered any collation which are multibytes? for example, Chinese?

File: test/python/expected/pyodbc/ddl_tables_index.out [hunk 6]
```diff
@@ -237,7 +237,7 @@ CREATE TABLE [dbo].[test_upper](
 	[a] [char](1) NULL
 ) ON [PRIMARY]
 
-ALTER TABLE [dbo].[test_upper]  WITH CHECK ADD  CONSTRAINT [test_upper_a_check] CHECK  (((upper(a) IN ('A', 'B'))))
+ALTER TABLE [dbo].[test_upper]  WITH CHECK ADD  CONSTRAINT [test_upper_a_check] CHECK  (((sys.upper(a) IN ('A', 'B'))))
```
Comment (Deepesh125) [inline test/python/expected/pyodbc/ddl_tables_index.out:240]: why is this needed? by default it would pick sys.upper, right?

File: test/JDBC/expected/BABEL-3395-vu-prepare.out (hunk 1 above)
Comment (Anikait143) [reply test/JDBC/expected/BABEL-3395-vu-prepare.out]: Yes, we have covered Chinese, Arabic, Japanese collation with multibyte characters.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 7]
```diff
@@ -3073,6 +3073,130 @@ CREATE OR REPLACE FUNCTION sys.substring(string sys.NCHAR, i INTEGER, j INTEGER)
...
+    RETURN (SELECT pg_catalog.upper(varch));
+END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE;
+
+-- Function to handle NCHAR because of return type NVARCHAR
+CREATE OR REPLACE FUNCTION sys.upper(sys.NCHAR)
+RETURNS sys.NVARCHAR
+AS $$
+BEGIN
+    RETURN (SELECT pg_catalog.upper($1));
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: NIT: SELECT is not needed

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 8]
```diff
@@ -3073,6 +3073,130 @@ CREATE OR REPLACE FUNCTION sys.substring(string sys.NCHAR, i INTEGER, j INTEGER)
...
+    varch sys.varchar;
+    typ_oid oid;
+    typnam text;
+BEGIN
+    typ_oid := (SELECT typbasetype FROM pg_type WHERE oid = pg_typeof($1));
+    typnam := (SELECT typname FROM pg_type WHERE oid in (typ_oid));
+    IF pg_typeof($1) IN ('image'::regtype, 'sql_variant'::regtype, 'xml'::regtype, 'geometry'::regtype, 'geography'::regtype) THEN
+        RAISE EXCEPTION 'Argument data type % is invalid for argument 1 of upper function.', pg_typeof($1);
+    ELSIF typnam IN ( 'image'::regtype::name COLLATE "C", 'sql_variant'::regtype::name COLLATE "C", 'xml'::regtype::name COLLATE "C", 'geometry'::regtype::name COLLATE "C", 'geography'::regtype::name COLLATE "C") THEN
+        RAISE EXCEPTION 'Argument data type % is invalid for argument 1 of upper function.', typnam;
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: Please use `translate_pg_type_to_tsql` function instead of SELECT queries to find the tsql typename.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 9]
```diff
@@ -3073,6 +3073,130 @@ CREATE OR REPLACE FUNCTION sys.substring(string sys.NCHAR, i INTEGER, j INTEGER)
...
+    END IF;
+    IF $1 IS NULL THEN
+        RETURN NULL;
+    END IF;
+    varch := (SELECT CAST ($1 AS sys.varchar));
+    -- Call the underlying function after preprocessing
+    RETURN (SELECT pg_catalog.upper(varch));
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:3098]: We don't need to explicit cast here, function return type will implicitly cast it.

**Review by sumitj824 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 7 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: Done, Removed Select from relevant places

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 8 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: Done, Optimized the code block

**Review by sumitj824 — APPROVED** [review]
LGTM

**Review by Deepesh125 — APPROVED** [review]
